src/main/java/hudson/plugins/git/GitSCM.java

```diff
@@ -970,22 +969,14 @@ public EnvVars getEnvironment() {
                                               final @NonNull GitClient git,
                                               final @NonNull TaskListener listener) throws IOException, InterruptedException {
         PrintStream log = listener.getLogger();
-        Collection<Revision> candidates = Collections.EMPTY_LIST;
+        Collection<Revision> candidates;
         final BuildChooserContext context = new BuildChooserContextImpl(build.getParent(), build, environment);
         getBuildChooser().prepareWorkingTree(git, listener, context);
 
-
-        // every MatrixRun should build the same marked commit ID
-        if (build instanceof MatrixRun) {
-            MatrixBuild parentBuild = ((MatrixRun) build).getParentBuild();
-            if (parentBuild != null) {
-                BuildData parentBuildData = getBuildData(parentBuild);
-                if (parentBuildData != null) {
-                    Build lastBuild = parentBuildData.lastBuild;
-                    if (lastBuild!=null)
-                        candidates = Collections.singleton(lastBuild.getMarked());
-                }
-            }
+        try {
+            candidates = GitSCMMatrixUtil.populateCandidatesFromMatrixBuild(build, this);
+        } catch (Throwable ex) { // in case there's no matrix-project plugin installed
```

Comment:
This should not be necessary. You can check for the `matrix-project` plugin being installed and enabled. Or, you can check the class name as you did in `GitPublisher`.